src/kernel.ts

```diff
@@ -3,1067 +3,793 @@
-            type: "POST",
-            contentType: false,  // there's no data with this
-            dataType: "json",
-            success: this._on_success(on_success),
-            error: this._on_error(on_error)
-        });
-    }
-
-    reconnect(): void {
-        /**
-         * Reconnect to a disconnected kernel. This is not actually a
-         * standard HTTP request, but useful function nonetheless for
-         * reconnecting to the kernel if the connection is somehow lost.
-         *
-         * @function reconnect
-         */
-        if (this.is_connected()) {
-            return;
-        }
-        this._reconnect_attempt = this._reconnect_attempt + 1;
-        events.trigger('kernel_reconnecting.Kernel', {
-            kernel: this,
-            attempt: this._reconnect_attempt,
-        });
-        this.start_channels();
-    }
-
-    _on_success(success): any {
-        /**
-         * Handle a successful AJAX request by updating the kernel id and
-         * name from the response, and then optionally calling a provided
-         * callback.
-         *
-         * @function _on_success
-         * @param {function} success - callback
-         */
-        return (data, status, xhr) => {
-            if (data) {
-                this.id = data.id;
-                this.name = data.name;
-            }
-            this.kernel_url = utils.url_join_encode(this.kernel_service_url, this.id);
-            if (success) {
-                success(data, status, xhr);
-            }
-        };
-    }
+/*
+ * Kernel Message specification.
+ */
+export
+  interface IKernelMsg {
+  header: IKernelMsgHeader;
+  metadata: any;
+  content: any;
+  parentHeader: any;
+  msgId?: string;
+  msgType?: string;
+  channel?: string;
+  buffers?: string[] | ArrayBuffer[];
+}
 
-    _on_error(error) {
-        /**
-         * Handle a failed AJAX request by logging the error message, and
-         * then optionally calling a provided callback.
-         *
-         * @function _on_error
-         * @param {function} error - callback
-         */
-        return (xhr, status, err) => {
-            utils.log_ajax_error(xhr, status, err);
-            if (error) {
-                error(xhr, status, err);
-            }
-        };
-    }
 
-    _kernel_created(data): void {
-        /**
-         * Perform necessary tasks once the kernel has been started,
-         * including actually connecting to the kernel.
-         *
-         * @function _kernel_created
-         * @param {Object} data - information about the kernel including id
-         */
-        this.id = data.id;
-        this.kernel_url = utils.url_join_encode(this.kernel_service_url, this.id);
-        this.start_channels();
-    }
+/**
+ * Object providing a Future interface for message callbacks.
+ */
+export
+interface IKernelFuture {
+  /**
+   * Dispose and unregister the future.
+   */   
+  dispose(): void;
+
+  /**
+   * Regsiter a reply handler. Returns 'this'.
```

Comment:
'this' -> `this` (use backquotes to format more nicely in Markdown).  Also below.